src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java

```diff
@@ -1,99 +1,256 @@
+            for(final SamPairUtil.PairOrientation orientation : rgStats.getValue().keySet()){
+                final SortedMap<Integer, Long> rgHist = rgStats.getValue().get(orientation);
+                histList.putIfAbsent(orientation, new TreeMap<>());
+                SortedMap<Integer, Long> hist = histList.get(orientation);
+                for(Map.Entry<Integer, Long> entry : rgHist.entrySet()){
+                    hist.putIfAbsent(entry.getKey(), 0L);
+                    Long binSize = hist.get(entry.getKey());
+                    binSize += entry.getValue();
+                    hist.put(entry.getKey(), binSize);
+                }
+            }
         }
-        return lengthArr;
+        return histogramsAtHigherLevel;
     }
 
-    private void collectMedianAndMAD(final JavaRDD<Integer> lengthRDD){
+    /**
+     * Last step worker function where hand-rolled histograms (SortedMap) is converted to htsjdk Histograms, and metrics information is collected
+     * @param mapping                hand-rolled histogram
+     * @param histogramMADTolerance  tolerance to trim histogram so "outliers" don't ruin mean and SD values
+     * @return                       htsjdk Histogram and InsertSizeMetrics bundled together under particular pair orientations, for the same grouping that's fed in
+     */
+    private static SortedMap<GroupInfo, Map<SamPairUtil.PairOrientation, Tuple2<Histogram<Integer>, InsertSizeMetrics>>> convertSortedMapToHTSHistogram(final SortedMap<GroupInfo, Map<SamPairUtil.PairOrientation, SortedMap<Integer, Long>>> mapping,
+                                                                                                                                                  final double histogramMADTolerance){
+
+        SortedMap<GroupInfo, Map<SamPairUtil.PairOrientation, Tuple2<Histogram<Integer>, InsertSizeMetrics>>> converted = new TreeMap<>(groupInfoComparator);
+
+        for(final Map.Entry<GroupInfo, Map<SamPairUtil.PairOrientation, SortedMap<Integer, Long>>> groupStats : mapping.entrySet()){
+
+            final GroupInfo groupInfo = groupStats.getKey();
 
-        double lengthArr[] = getRDDArray(lengthRDD);
+            for(final Map.Entry<SamPairUtil.PairOrientation, SortedMap<Integer, Long>> groupStatsOfAOrientation: groupStats.getValue().entrySet()){
 
-        final DescriptiveStatistics percent = new DescriptiveStatistics(lengthArr);
+                final SamPairUtil.PairOrientation orientation = groupStatsOfAOrientation.getKey();
+                // convert to htsjdk Histogram
+                final Histogram<Integer> htsHist = new Histogram<>("insert_size", getGroupName(groupInfo) + orientationToString(orientation) + "_count");
+                final SortedMap<Integer, Long> hist = groupStatsOfAOrientation.getValue();
+                for(final int size : hist.keySet()){
+                    htsHist.prefillBins(size);
```

Comment:
Do you need to do this prefill step? It looks like you might not, reading the code.